File: 0200-number-of-islands/0200-number-of-islands.go
// var dirs = [][2]int{{-1, 0}, {0, 1}, {1, 0}, {0, -1}}

func numIslands(grid [][]byte) int {
    n, m := len(grid), len(grid[0])


    visited := make([][]bool, n)

    for i := range n {
        visited[i] = make([]bool, m)
    }

    var dfs func (row, col int)

    dfs = func (row, col int) {
        if !(row >= 0 && row < n && col >= 0 && col < m) || 
        grid[row][col] == '0' || 
        visited[row][col] {
            return
        }

        visited[row][col] = true

        dfs(row-1, col) // up
        dfs(row, col+1) // right
        dfs(row+1, col) // down
        dfs(row, col-1) // left

    }

    islands := 0
    for r := range n {
        for c := range m {
            if grid[r][c] == '0' || visited[r][c] {
                continue
            }
            dfs(r, c)
            islands++
        }
    }

    return islands
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n, m = len(grid), len(grid[0])


        visited = [[0] * m for _ in range(n)]
        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

        def dfs(r, c):
            visited[r][c] = 1

            for dir_r, dir_c in directions:
                new_r, new_c = dir_r + r, dir_c + c

                if 0 <= new_r < n and 0 <= new_c < m and grid[new_r][new_c] == "1" and not visited[new_r][new_c]:
                    dfs(new_r, new_c)
        

        islands = 0
        for row in range(n):
            for col in range(m):
                if grid[row][col] == "1" and not visited[row][col]:
                    dfs(row, col)
                    islands += 1
        
        return islands